cherry_picker/cherry_picker.py

```diff
@@ -296,11 +296,31 @@ def get_updated_commit_message(self, cherry_pick_branch):
         commit_prefix = ""
         if self.prefix_commit:
             commit_prefix = f"[{get_base_branch(cherry_pick_branch)}] "
-        return f"""{commit_prefix}{self.get_commit_message(self.commit_sha1)}
-(cherry picked from commit {self.commit_sha1})
+        updated_commit_message = f"{commit_prefix}{self.get_commit_message(self.commit_sha1)}"
+        cherry_pick_information = f"(cherry picked from commit {self.commit_sha1})\n:"
+        cmd = [
+            "git",
+            "interpret-trailers",
+            "--where",
+            "start",
+            "--trailer",
+            f"Co-authored-by: {get_author_info_from_short_sha(self.commit_sha1)}",
+            "--trailer",
+            cherry_pick_information,
+        ]
+        output = subprocess.check_output(cmd, input=updated_commit_message.encode())
```

Comment:
(One of the comments from my previous review got lost.)

It would be helpful to add a docstring and a couple more comments to this function to explain what it does, since it's not immediately obvious.
AFAIK and by looking at the tests, it seems it retrieves the original commit message and:
* adds a prefix at the beginning with the branch the commit is being backported to
* adds a line in the middle mentioning the original commit id
* adds the co-authors (and possibly other metadata?) at the end

I also wonder if there's a more straightforward way to do achieve this.
Can you use `git interpret-trailers --only-trailers` to extract the metadata and then do something like this?
```py
f"""{commit_prefix}{original_commit_message}
(cherry picked from commit {self.commit_sha1})

{metadata}"""
```

Reply:
Welp, some value sure was lost by me not adding more comments in there in the first place... I added some now to explain the way this is supposed to work.

> I also wonder if there's a more straightforward way to do achieve this. Can you use `git interpret-trailers --only-trailers` to extract the metadata and then do something like this?
> 
> ```python
> f"""{commit_prefix}{original_commit_message}
> (cherry picked from commit {self.commit_sha1})
> 
> {metadata}"""
> ```

I can look into an alternative way of doing this. The important things are:
- the trailers cannot break after the addition of `(cherry picked from commit ...)`
- adding Co-authored-by should not cause duplicates
- the order of the trailers probably shouldn't be affected more than needed by the changes we make (currently we only insert a new trailer at the beginning of existing trailers and don't touch them further)

One of the problems is that `git interpret-trailers` does not have a flag that does the opposite of `--only-trailers` so getting the "original commit message" may not be as easy as it seems. A naive way of doing it would be to cut the end of the full message by the length of the output of `git interpret-trailers` but I wouldn't be too surprised if this wouldn't work that well in practice (as I said at the beginning, some value sure was lost by me not writing some more comments when I wrote this PR so I'll have to look into that).

Reply:
> adding Co-authored-by should not cause duplicates

If we can isolate all the trailers (and `--only-trailers` seems to do that) you could pass them through a `set()` to remove duplicates.  Even if the order is lost I'm not sure it matters.  If you want to preserve it, it's pretty easy to write a function that removes duplicates without losing the order.

> getting the "original commit message" may not be as easy as it seems.

Recently I was looking into manually adding co-authors, and apparently you had to leave two empty lines between the message and the trailers, so an `msg = full_commit_msg.rsplit('\n\n', 1)[0]` might work (assuming this is enforced).  Otherwise we can cut the end like you said or remove individual trailer lines from `full_commit_msg.splitlines()`.

Reply:
> If we can isolate all the trailers (and `--only-trailers` seems to do that) you could pass them through a `set()` to remove duplicates. Even if the order is lost I'm not sure it matters. If you want to preserve it, it's pretty easy to write a function that removes duplicates without losing the order.

cherry-picker is Python 3.7+ so this could just rely on the insertion order of dictionaries, possibly with the help of `dict.fromkeys()`.

> Recently I was looking into manually adding co-authors, and apparently you had to leave two empty lines between the message and the trailers, so an `msg = full_commit_msg.rsplit('\n\n', 1)[0]` might work (assuming this is enforced).

The problem with this is that we would need to write a proper detection of trailers or we risk the trailers no longer being trailers. A single split is not enough as we plan on adding a trailer and that means we need to make sure that we add a trailer with other trailers. For example, if this were the original commit message:

```
Commit message

some text
```

Then we can't just simply add it to the last part:

```
Commit message

some text
Co-authored-by: Author Name <author@name.email>
```
On the other hand, we also have to keep all trailers together so if this were the original commit message:

```
Commit message

Co-authored-by: Co-Author Name <co@author.email>
```

Then this would not be acceptable:

```
Commit message

Co-authored-by: Co-Author Name <co@author.email>

Co-authored-by: Author Name <author@name.email>
```

For reference, here's the logic that git uses: https://github.com/git/git/blob/afa70145a25e81faa685dc0b465e52b45d2444bd/trailer.c#L818-L915

> Otherwise we can cut the end like you said

Based on these two:
https://github.com/git/git/blob/afa70145a25e81faa685dc0b465e52b45d2444bd/trailer.c#L1108-L1123
https://github.com/git/git/blob/afa70145a25e81faa685dc0b465e52b45d2444bd/trailer.c#L982-L990
*I think* there's no alteration done to the printed trailers which would make this doable but I'll have to read into it just a bit more to be sure. I didn't want to delay my response further since I might not be able to get to it at all today so I have not yet carefully examined these two linked fragments of code.

> or remove individual trailer lines from `full_commit_msg.splitlines()`.

I think this falls into same category as `rsplit` but I'm not sure I fully understand what you mean by this.

Reply:
> I'm not sure I fully understand what you mean by this.

Given this:
```
$ cat msg.txt 
gh-12345: fix some broken stuff

Some stuff was broken and now
it's fixed.

Co-authored-by: Co-Author Name <co@author.email>
Co-authored-by: Author Name <author@name.email>
```
```sh
$ cat msg.txt | git interpret-trailers --only-trailers 
Co-authored-by: Co-Author Name <co@author.email>
Co-authored-by: Author Name <author@name.email>
```
You could do something like this:
```pycon
>>> # get these from self.get_commit_message(self.commit_sha1).splitlines()
>>> with open('msg.txt') as f:
...   lines = [line.strip() for line in f]  # remove trailing whitespace/newlines
... 
>>> # get these from `git interpret-trailers --only-trailers`
>>> trailers = ['Co-authored-by: Co-Author Name <co@author.email>',
...             'Co-authored-by: Author Name <author@name.email>']
>>> # remove trailers from the full message
>>> for trailer in trailers:
...   lines.remove(trailer)
... 
>>> msg = '\n'.join(lines).strip()  # this is the msg without trailers
>>> trailers.append('Co-authored-by: Another Author')
>>> metadata = '\n'.join(trailers)  # these are the updated trailers
>>> print(f"""[3.10] {msg}
... (cherry-picked from somewhere)
... 
... {metadata}""")  # this is the final message
[3.10] gh-12345: fix some broken stuff

Some stuff was broken and now
it's fixed.
(cherry-picked from somewhere)

Co-authored-by: Co-Author Name <co@author.email>
Co-authored-by: Author Name <author@name.email>
Co-authored-by: Another Author
```
That is, if you take the full commit and remove the trailer lines you obtained with `git interpret-trailers --only-trailers` you'll be left with the message only.  Once you have the message and the trailers you can update the trailers and recombine everything in a new msg.  Do you think this will work?